Fix custom catalog store plugin loading (v2)

hashhar opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Currently a custom catalog store plugin cannot be used because the Guice bindings are incorrect.

Fixes https://github.com/trinodb/trino/issues/22554
Supercedes https://github.com/trinodb/trino/pull/22784

I couldn't think of any way to test this other than a product test - but that too would need me to add a custom plugin to the build just to test this - I manually tested this instead for now.

<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

See https://github.com/trinodb/trino/issues/22554 for details. And https://github.com/trinodb/trino/pull/20489 for earlier changes.

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# General
* Fix loading of custom `CatalogStore` implementations. Set `catalog.store` to the name of the custom `CatalogStoreFactory`. `catalog-store.name` in `etc/catalog-store.properties` is no longer required. ({issue}`22554`)
```

dain commented on core/trino-main/src/main/java/io/trino/connector/DynamicCatalogManagerModule.java:
```diff
@@ -35,13 +37,15 @@ protected void setup(Binder binder)
                     binder.bind(CatalogStore.class).to(FileCatalogStore.class).in(Scopes.SINGLETON);
                 }
                 default -> {
+                    newOptionalBinder(binder, Key.get(String.class, CatalogStoreKind.class)).setBinding().toInstance(catalogStoreKind);
```
I don't think the optional binder is needed (is more complex).  CatalogStoreManager can just have a dependency on CatalogStoreConfig.

hashhar replied:
🤦 I was thinking there must be a simpler way but was tunnel-visioned at the time.

hashhar replied:
Resolved.

hashhar commented on core/trino-main/src/main/java/io/trino/connector/CatalogStoreKind.java:
```diff
@@ -0,0 +1,31 @@
+@BindingAnnotation
+public @interface CatalogStoreKind
+{
+}
```
This is now gone since we use the config class instead of binding a string.

hashhar commented:
Addressed your comments Dain. PTAL.

I'll send a separate PR that moves file + memory catalog store to be loaded using the `CatalogStoreManager` too instead of being directly bound like we discussed offline.

hashhar commented:
Forgot to tag you @dain. 😄

sebadiaz commented on core/trino-main/src/test/java/io/trino/server/testing/TestTestingTrinoServer.java:
```diff
@@ -63,4 +64,26 @@ void testSetCatalogManagementToStatic()
+            assertThat(server.getInstance(Key.get(CatalogStore.class)))
+                    .isInstanceOf(FileCatalogStore.class);
+        }
+    }
```
You should test the Dynamic custom store. 
In invoking a New plugin based on catalogstore catalogstorefactory and plugin interface that contains the getcatalogstorefactoriesn method.

My last test in your code configuration let me confuse. I m not sure that it works well on a custom plugin for catalog store.

sebadiaz replied:
Thanks
